examples/vision/image_classification_with_vision_transformer.py

```diff
@@ -0,0 +1,294 @@
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/30
+Last modified: 2021/01/30
+Description: Implementing Vision Transformer (ViT) model for image classification.
+"""
+
+"""
+## Introduction
+
+This example implements the [Vision Transformer (ViT)](https://arxiv.org/abs/2010.11929)
+model, by Alexey Dosovitskiy et al., for CIFAR-100 image classification.
+The ViT applies the transformer architecture with self-attentions to sequences of
+image patches, without using convolutional networks.
+
+The example requires TensorFlow 2.4 or higher, and
+[TensorFlow Addons](https://www.tensorflow.org/addons/overview),
+which can be installed using the following command:
+
+```python
+pip install -U tensorflow-addons
+```
+"""
+
+"""
+## Setup
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_addons as tfa
+
+"""
+## Prepare the data
+"""
+
+num_classes = 100
+input_shape = (32, 32, 3)
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar100.load_data()
+
+print(f"x_train shape: {x_train.shape} - y_train shape: {y_train.shape}")
+print(f"x_test shape: {x_test.shape} - y_test shape: {y_test.shape}")
+
+"""
+## Define hyperparameters
+"""
+
+learning_rate = 0.003
+weight_decay = 0.001
+batch_size = 512
+hidden_units = [128]
+num_epochs = 100
+dropout_rate = 0.5
+
+"""
+## Compile, train, and evaluate the mode
+"""
+
+
+def run_experiment(model):
+    k = 5
+    model.compile(
+        optimizer=tfa.optimizers.AdamW(
+            learning_rate=learning_rate, weight_decay=weight_decay
+        ),
+        loss=keras.losses.SparseCategoricalCrossentropy(),
+        metrics=[keras.metrics.SparseTopKCategoricalAccuracy(k)],
+    )
+
+    history = model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=num_epochs)
+
+    accuracy = model.evaluate(x_test, y_test)[1]
+    print(f"Test tok {k} accuracy: {round(accuracy * 100, 2)}%")
+
+    return history
+
+
+"""
+## Using data augmentation
+"""
+
+data_augmentation = keras.Sequential(
+    [
+        layers.experimental.preprocessing.Rescaling(1.0 / 255),
+        layers.experimental.preprocessing.RandomFlip("horizontal"),
+        layers.experimental.preprocessing.RandomRotation(0.02),
+    ],
+    name="data_augmentation",
+)
+
+"""
+## Implement Multilayer perceptron (MLP) as a layer
+"""
+
+
+class MLP(layers.Layer):
+    def __init__(self, hidden_units, dropout_rate):
+        super(MLP, self).__init__()
+        mlp_layers = []
+        for units in hidden_units:
+            mlp_layers.append(layers.Dense(units, activation=tf.nn.gelu))
+            mlp_layers.append(layers.Dropout(dropout_rate))
+        self.mlp = keras.Sequential(mlp_layers)
+
+    def call(self, inputs):
+        return self.mlp(inputs)
+
+
+"""
+## Experiment 1: Train the baseline classification model
+
+We use an untrained ResNet50 architecture as our baseline model.
```

Comment:
Suggestion: ResNet-50
```suggestion
We use an untrained ResNet-50 architecture as our baseline model.
```

Reply:
Done